File: 1]. DSA + CP/1]. DSA/1]. Data Structures/01]. Array/C++/_008)_Left_Rotate_Array_by_d_(Time_N).cpp
#include <iostream>
using namespace std;

void reverse(int arr[], int low, int high)
{
	while(low < high)
	{
		swap(arr[high], arr[low]);
		low++;
		high--;
	}
}

void leftRotate(int arr[], int d, int n)
{
	reverse(arr, 0, d - 1);
	reverse(arr, d, n - 1);
	reverse(arr, 0, n - 1);
}

int main()
{
      int arr[] = {1, 2, 3, 4, 5}, n = 5, d = 4;
      cout<<"Before Left Rotation"<<endl;
       for(int i = 0; i < n; i++)
       {
       		cout<<arr[i]<<" ";
       }
       cout<<endl;
       leftRotate(arr, d, n);
       cout<<"After '4' Left Rotation"<<endl;
       for(int i = 0; i < n; i++)
       {
       		cout<<arr[i]<<" ";
       }
}

File: 1]. DSA + CP/1]. DSA/1]. Data Structures/01]. Array/Dart/_008)_Left_Rotate_Array_by_d_(Time_N).dart
void reverse(arr, low, high) {
  while (low < high) {
    var temp = arr[low];
    arr[low] = arr[high];
    arr[high] = temp;

    low++;
    high--;
  }
}

void leftRotate(arr, d, n) {
  reverse(arr, 0, d - 1);
  reverse(arr, d, n - 1);
  reverse(arr, 0, n - 1);
}

void main() {
  var arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
  print('Before Left Rotate: ${arr}');
  leftRotate(arr, 4, arr.length);
  print('After Left Rotate: ${arr}');
}
